### PR TITLE
Ksp 1.2 update (WIP)

### PR DESCRIPTION
**Pull request by alexanderabramov**
@ferram4 I haven't seen any branches with work on 1.2 update, so I've started some work on it.  Please let me know if you have any WIP as well

### CODE REVIEW

File: FerramAerospaceResearch/LEGACYferram4/FARWingAerodynamicModel.cs [hunk 1]
```diff
@@ -452,26 +452,27 @@ public void StartInitialization()
...
-            if (part is Winglet)
-            {
-                Winglet w = part as Winglet;
-                w.deflectionLiftCoeff = 0;
-                w.dragCoeff = 0;
-                w.angularDrag = 0;
-                w.maximum_drag = 0;
-                w.minimum_drag = 0;
-            }
+            // TODO 1.2: verify if these are renamed or really gone
```
Comment (ayan4m1) [inline FerramAerospaceResearch/LEGACYferram4/FARWingAerodynamicModel.cs]: Do a find here for "ControlSurface" - https://kerbalspaceprogram.com/api/classes.html

So there is a `ModuleControlSurface` but I don't see any reference to Wing/winglet anymore.

Comment (ferram4) [inline FerramAerospaceResearch/LEGACYferram4/FARWingAerodynamicModel.cs]: Yep, they're gone.  1.2 had a lot of code cleanup.

**Comment by ferram4** [comment]
Hey, thanks for getting to work on this.  I'll look it over and see about merging it into a dev build branch.

Outside of the KSP API fixes the thing that needs to be checked is its behavior with ModularFlightIntegrator due to a lot of changes in the stock FlightIntegrator from 1.1.3 -> 1.2.  Make sure you're developing with MFI 1.2.2 and be very wary of what happens since there is a chance of things going out-of-order as a consequence of changes.

**Comment by alexanderabramov** [comment]
Thanks for the heads up, I'm using ModularFlightIntegrator compiled from master.

By the way, have you found a good way to debug multi-threaded issues with VS and Unity? I'm attached in the debugger, but when KSP deadlocks VS cannot do anything about it, I guess that funky remote debugger hangs as well.

**Comment by alexanderabramov** [comment]
@ferram4 Do you have the source code for farshaders.ksp? I get this error message in logs and cross-section graph is painted pink/purple: "GpuProgram creation error: shader program type is unrecognised. You might have a precompiled shader asset from an old Unity version." If I comment out the custom shader, everything works.

File: FerramAerospaceResearch/FARThreading/VoxelizationThreadpool.cs [hunk 2]
```diff
@@ -137,36 +137,26 @@ public void QueueVoxelization(Action voxelAction)
 
         public void RunOnMainThread(Action action)
         {
-            if (RunInMainThread)
```
Comment (ferram4) [inline FerramAerospaceResearch/FARThreading/VoxelizationThreadpool.cs]: This is to handle the debugger that everyone besides me seems to use that involves the UnityEditor.  It throws horrible fits if stuff runs off the main thread, but KSP itself is perfectly fine with things running in their own thread.

I wouldn't know the specifics of what happens here, as I'm not the one who made the changes to this or to KSP.

**Comment by ferram4** [comment]
I have the source sitting around, don't worry, I'll take care of that.

**Comment by ferram4** [comment]
Yo, sorry about the delay @alexanderabramov.

I've merged what's here onto the ferram4:KSP_update branch for development.  Please check it to make sure that I haven't somehow broken anything bringing it over.  I'll start going over it to make sure that everything works fine and make sure that there aren't any weird new physics behaviors in there.  Once everything is hunky dory I'll merge that into master for a full release.

**Comment by Fl4shblade** [comment]
Hello @ferram4 ,

I downloaded this branch and so far it seems to be running very stable already. I just came across a couple of small issues so far. The transsonic curves are all purple color. The FAR buttons are in scenes where they are not suppossed to be. The VAB/SPH button remains in the flight scene and the flight scene button remains in the VAB/SPH leaving a none functional button. I haven't discovered any weirdness with the heart of the matter, the aerodynamics themselves. All seems to be good on this front.

I doubt you are not aware of these issues but I still thought reporting couldn't hurt. Hope it helps and thanks for FARs existence.

**Comment by Fl4shblade** [comment]
Hi ferram4 here is my preliminary report on your update to the branch. As you probably already know the transonic graphs work again in windows. As for the FAR buttons. That has mixed results. I have only one button in the VAB and SPH. When I start the game and go directly to the runway and launch I also only have the active FAR button. When I revert to the SPH the flight scene button remains so 2 buttons. When I start the game, go the the SPH pick a craft and launch it the SPH button remains in the list, 2 buttons again. A first is that after recovering the vessel and leaving the SPH there is 2 FAR buttons in the main KSP overview now. Going back to the start menu also 2 FAR buttons remain which is another first. When I load my save again there are now three buttons in the KSP overview. Klicking on the SPH one button gets destroyed one remains, 2 buttons. Also 2 buttons in the flight scene when I load a vessel and launch. At one point I had three FAR buttons in the flight scene but I fail at recreating the condition. 

The most important thing, the aerodynamics still doesn't give me any trouble.

**Comment by ferram4** [comment]
@Fl4shblade The multiple toolbar buttons are being investigated and will be fixed.
The phantom forces are due to an out-of-date version of KJR.  Update it.

**Comment by Fl4shblade** [comment]
@ferram4 Oh my I indeed missed a KJR release there. Everything works fine with the latest release. My apologies.

**Comment by ferram4** [comment]
Very strange, this should be listed as merged after it was merged onto ksp_update and then that was merged into master, but it hasn't.  I'll be closing this then just to keep the PRs clear (I don't dare merge and risk disaster), but rest assured that the commits have been merged, and thanks for the work on it.

**Pull request closed without merging**